transformer_lens/HookedSAETransformer.py

```diff
@@ -0,0 +1,271 @@
+            Float[torch.Tensor, "batch pos d_vocab"],
+            Loss,
+            Tuple[Float[torch.Tensor, "batch pos d_vocab"], Loss],
+        ],
+        Union[ActivationCache, Dict[str, torch.Tensor]],
+    ]:
+        """Wrapper around 'run_with_cache' in HookedTransformer.
+
+        Turns on the SAEs for the given act_names before running the model with cache and then turns them off after
+        Note this will turn off all other SAEs that are not in act_names before running
+        After running, it will turn off all SAEs
+
+        Args:
+            *model_args: Positional arguments for the model forward pass
+            act_names: (List[str]) The act_names for the SAEs to turn on for this forward pass
+            return_cache_object: (bool) if True, this will return an ActivationCache object, with a bunch of
+                useful HookedTransformer specific methods, otherwise it will return a dictionary of
+                activations as in HookedRootModule.
+            remove_batch_dim: (bool) Whether to remove the batch dimension (only works for batch_size==1). Defaults to False.
+            **kwargs: Keyword arguments for the model forward pass
+        """
+        self.turn_saes_off()
+        try:
+            self.turn_saes_on(act_names)
+            out = self.run_with_cache(
+                *model_args,
+                return_cache_object=return_cache_object,
+                remove_batch_dim=remove_batch_dim,
+                **kwargs,
+            )
+        finally:
+            self.turn_saes_off()
+        return out
+
+    def run_with_hooks_with_saes(
+        self,
+        *model_args,
+        act_names: List[str] = [],
+        fwd_hooks: List[Tuple[Union[str, Callable], Callable]] = [],
+        bwd_hooks: List[Tuple[Union[str, Callable], Callable]] = [],
+        reset_hooks_end=True,
+        clear_contexts=False,
+        **model_kwargs,
+    ):
+        """Wrapper around 'run_with_hooks' in HookedTransformer.
+
+        Turns on the SAEs for the given act_names before running the model with hooks and then turns them off after
+        Note this will turn off all other SAEs that are not in act_names before running
+        After running, it will turn off all SAEs
+
+        ARgs:
+            *model_args: Positional arguments for the model forward pass
+            act_names: (List[str]) The act_names for the SAEs to turn on for this forward pass
+            fwd_hooks: (List[Tuple[Union[str, Callable], Callable]]) List of forward hooks to apply
+            bwd_hooks: (List[Tuple[Union[str, Callable], Callable]]) List of backward hooks to apply
+            reset_hooks_end: (bool) Whether to reset the hooks at the end of the forward pass (default: True)
+            clear_contexts: (bool) Whether to clear the contexts at the end of the forward pass (default: False)
+            **model_kwargs: Keyword arguments for the model forward pass
+        """
+        self.turn_saes_off()
```

Comment:
IMO the main reason to use run_with_saes is if you just want to run with the specific SAEs you pass in for one forward pass, then have a clean slate. I can't imagine people using it in addition to other SAEs they already have turned on, unless by mistake. What do you think?

Reply:
that makes sense to me.

Reply:
I like the idea of a flag, and think it should be by default off. I think that the most important thing here is that the API is intuitive and adding the flag makes this more obvious. FWIW, I share Connor's intuition about how it will be used but can imagine wanting to run with *marginal saes* in some cases (adding / removing SAEs on the margin and not having a reset). This seems more likely in a world with great SAEs (way better than we have now).